openpype/hosts/maya/startup/userSetup.py

```diff
@@ -1,16 +1,67 @@
 import os
+import shutil
+
 from openpype.settings import get_project_settings
 from openpype.pipeline import install_host
-from openpype.hosts.maya.api import MayaHost
+from openpype.hosts.maya.api import MayaHost, current_file
+from openpype.lib import register_event_callback
+
 from maya import cmds
 
 host = MayaHost()
 install_host(host)
 
 
-print("starting OpenPype usersetup")
+print("Starting OpenPype usersetup...")
+
+# Open Workfile Post Initialization.
+key = "OPENPYPE_OPEN_WORKFILE_POST_INITIALIZATION"
+if bool(int(os.environ.get(key, "0"))):
+    cmds.evalDeferred(
+        "cmds.file(os.environ[\"AVALON_LAST_WORKFILE\"], open=True,"
+        " force=True)",
+        lowestPriority=True
+    )
+
+
+# Setup Xgen save callback.
+def xgen_on_save():
+    """Increments the xgen side car files .xgen and .xgd
+
+    Only works when incrementing to the same directory.
+    """
+
+    file_path = current_file()
+    current_dir = os.path.dirname(file_path)
+    basename = os.path.basename(file_path).split(".")[0]
+    attrs = ["xgFileName", "xgBaseFile"]
+    for palette in cmds.ls(type="xgmPalette"):
+        for attr in attrs:
+            source = os.path.join(
+                current_dir, cmds.getAttr(palette + "." + attr)
+            )
+            if not os.path.exists(source):
+                continue
+
+            destination_basename = "{}__{}{}".format(
+                basename,
+                palette.replace(":", "_"),
+                os.path.splitext(source)[1]
+            )
+            destination = os.path.join(current_dir, destination_basename)
+
+            if source == destination:
+                continue
+
+            shutil.copy(source, destination)
+            cmds.setAttr(
+                palette + "." + attr, destination_basename, type="string"
+            )
+
+
+register_event_callback("save", xgen_on_save)
```

Comment:
I feel like this logic shouldn't be directly in `userSetup.py` but instead be somewhere in `pipeline.py` or at least somewhere where it gets triggered through OpenPype host install. No?

Looking at most host implementations the actual 'startup.py' files remain rather minimal and just trigger the host install, which then sets up the rest.

Reply:
Have removed the `xgen_on_save` function now. Do you still think we should move the `OPENPYPE_OPEN_WORKFILE_POST_INITIALIZATION` code?